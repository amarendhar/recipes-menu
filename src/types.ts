export type Maybe<T> = T | null;

export type Recipe = {
  id: string;
  title: string;
  image: string;
  tags: string[];
  description: string;
  chefName: string;
};
