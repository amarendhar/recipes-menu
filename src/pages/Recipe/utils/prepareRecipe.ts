import { EntryFields, EntrySkeletonType } from "contentful";
import { Maybe, RecipeEntry, RecipeData } from "types";

type RecipeFields = {
  title: EntryFields.Text;
  photo: EntryFields.AssetLink;
  tags: EntryFields.Array<EntryFields.EntryLink<EntrySkeletonType>>;
  description: EntryFields.Text;
  chef: EntryFields.EntryLink<EntrySkeletonType>;
};

export const prepareRecipe = (data: Maybe<RecipeEntry>): Maybe<RecipeData> => {
  if (!data) {
    return null;
  }

  const fields = data?.fields as RecipeFields;

  const id: string = data.sys.id;
  const title = fields?.title || "";
  const image = (fields?.photo?.fields?.file?.url || "") as string;
  const tags =
    fields?.tags?.map((tag) => (tag?.fields?.name as string) || "") || [];
  const description = fields?.description || "";
  const chef = (fields?.chef?.fields?.name || "") as string;
  const rating = 0;

  const recipeData: RecipeData = {
    id,
    title,
    image,
    tags,
    description,
    chef,
    rating,
  };

  return recipeData;
};
