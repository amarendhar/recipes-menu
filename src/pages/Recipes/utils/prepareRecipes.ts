import { Entry, EntryFields, EntrySkeletonType } from "contentful";
import { Maybe, RecipesEntry, RecipesItemData } from "types";

type RecipeFields = {
  title: EntryFields.Text;
  photo: EntryFields.AssetLink;
  tags: EntryFields.Array<EntryFields.EntryLink<EntrySkeletonType>>;
  description: EntryFields.Text;
  chef: EntryFields.EntryLink<EntrySkeletonType>;
};

export const prepareRecipes = (
  data: Maybe<RecipesEntry>
): RecipesItemData[] => {
  if (!data?.items) {
    return [];
  }

  const recipes: RecipesItemData[] = data.items.map(
    (recipe: Entry<EntrySkeletonType, undefined, string>) => {
      const fields = recipe?.fields as RecipeFields;

      const id: string = recipe.sys.id;
      const title = fields?.title || "";
      const image = (fields?.photo?.fields?.file?.url || "") as string;

      const recipeData: RecipesItemData = {
        id,
        title,
        image,
      };

      return recipeData;
    }
  );

  return recipes;
};
