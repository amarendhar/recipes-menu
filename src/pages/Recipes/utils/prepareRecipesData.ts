import {
  Entry,
  EntryFields,
  EntryCollection,
  EntrySkeletonType,
} from "contentful";
import { Maybe, Recipe } from "types";

type RecipeFields = {
  title: EntryFields.Text;
  photo: EntryFields.AssetLink;
  tags: EntryFields.Array<EntryFields.EntryLink<EntrySkeletonType>>;
  description: EntryFields.Text;
  chef: EntryFields.EntryLink<EntrySkeletonType>;
};

export const prepareRecipes = (
  data: Maybe<EntryCollection<EntrySkeletonType, undefined, string>>
): Recipe[] => {
  if (!data?.items) {
    return [];
  }

  const recipes: Recipe[] = data.items.map(
    (recipe: Entry<EntrySkeletonType, undefined, string>) => {
      const fields = recipe?.fields as RecipeFields;

      const id: string = recipe.sys.id;
      const title = fields?.title || "";
      const image = (fields?.photo?.fields?.file?.url || "") as string;
      const tags = fields?.tags?.map(
        (tag) => (tag?.fields?.name as string) || ""
      );
      const description = fields?.description || "";
      const chefName = (fields?.chef?.fields?.name || "") as string;

      const recipeData: Recipe = {
        id,
        title,
        image,
        tags,
        description,
        chefName,
      };

      return recipeData;
    }
  );

  return recipes;
};
