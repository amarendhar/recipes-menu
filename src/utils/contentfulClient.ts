import { createClient, CreateClientParams } from "contentful";
import {
  CONTENTFUL_SPACE_ID,
  CONTENTFUL_ACCESS_TOKEN,
  CONTENTFUL_ENVIRONMENT_ID,
} from "globalConstants";

const clientParams: CreateClientParams = {
  space: CONTENTFUL_SPACE_ID,
  accessToken: CONTENTFUL_ACCESS_TOKEN,
};

if (CONTENTFUL_ENVIRONMENT_ID) {
  clientParams.environment = CONTENTFUL_ENVIRONMENT_ID;
}

export const client = createClient(clientParams);
