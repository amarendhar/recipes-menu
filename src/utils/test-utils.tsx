import { Route, Routes, BrowserRouter } from "react-router-dom";
import {
  render as rtlRender,
  renderHook as rtlRenderHook,
} from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { defaultTheme } from "themes";
import { ContentfulProvider } from "api";
import App from "App";

const render = (
  ui: React.ReactElement,
  { route = "/" }: { route?: string | null } = {}
) => {
  window.history.pushState({}, "Test page", route);

  const Wrapper = ({ children }: { children: React.ReactNode }) =>
    route ? (
      <App />
    ) : (
      <ContentfulProvider>
        <ThemeProvider theme={defaultTheme}>
          <BrowserRouter>{children}</BrowserRouter>
        </ThemeProvider>
      </ContentfulProvider>
    );

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  };
};

const renderHook = (
  hook: any,
  { route = "/" }: { route?: string | null } = {}
) => {
  window.history.pushState({}, "Test page", route);

  const Wrapper = ({ children }: { children: any }) => (
    <ContentfulProvider>
      <BrowserRouter>
        <Routes>
          <Route path="/" element={children} />
          <Route path="/recipe/:id" element={children} />
        </Routes>
      </BrowserRouter>
    </ContentfulProvider>
  );

  return rtlRenderHook(() => hook(), { wrapper: Wrapper });
};

export * from "@testing-library/react";
export { render, renderHook };
