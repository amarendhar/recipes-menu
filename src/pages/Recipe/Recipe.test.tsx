import { render, screen, within, waitFor } from "utils/test-utils";
import { render as rtlRender } from "@testing-library/react";
import { compiler } from "markdown-to-jsx";
import { Recipe } from ".";
import { prepareRecipe } from "./utils";
import { RecipeEntry } from "types";
import mockRecipes from "mocks/__fixtures__/mockRecipes.json";

describe("# Recipe", () => {
  const renderComponent = ({
    route = `/recipe/${mockRecipes.items[0].sys.id}`,
  } = {}) => render(<Recipe />, { route });

  it("Should render Loading-Component, when fetching recipe", async () => {
    renderComponent();

    screen.getByTestId("recipe-loading");
    expect(screen.queryByTestId("recipe-error")).not.toBeInTheDocument();
    expect(screen.queryByTestId("recipe-container")).not.toBeInTheDocument();
  });

  it("Should fetch and render recipe with details image, title, description, tags and chef", async () => {
    const recipe = prepareRecipe(mockRecipes.items[3] as RecipeEntry);
    const { container } = renderComponent({
      route: `/recipe/${recipe?.id}`,
    });

    await waitFor(() => {
      screen.getByTestId("recipe-content");
    });

    expect(
      (screen.getByTestId("recipe-image") as HTMLImageElement).src
    ).toContain(recipe?.image);
    expect(
      (screen.getByTestId("recipe-image") as HTMLImageElement).alt
    ).toEqual(recipe?.title);
    expect(screen.getByTestId("recipe-title").textContent).toEqual(
      recipe?.title
    );
    expect(screen.getByTestId("recipe-description").textContent).toEqual(
      rtlRender(compiler(recipe?.description as string)).container.textContent
    );

    const recipeTagsContainer = screen.getByTestId("recipe-tags-container");
    within(recipeTagsContainer).getByText("Tags:");

    const recipeTags = screen.getAllByTestId("recipe-tag");
    recipe?.tags?.forEach((tagName, key) => {
      const recipeTag = recipeTags[key];
      expect(recipeTag.textContent).toEqual(tagName);
    });

    expect(screen.getByTestId("recipe-chef").textContent).toEqual(
      `Chef:${recipe?.chef}`
    );

    expect(container).toMatchSnapshot();
  });
});
