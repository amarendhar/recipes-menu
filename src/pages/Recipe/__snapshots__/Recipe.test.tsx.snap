// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# Recipe Should fetch and render recipe with details image, title, description, tags and chef 1`] = `
<div>
  <header
    class="sc-bczSft iRgQoa"
  >
    <div
      class="sc-gsnScm gKnCNt"
    >
      <a
        class="sc-dkzCjb jEMETj"
        href="/"
      >
        <h2>
          Recipes Menu
        </h2>
      </a>
    </div>
  </header>
  <main
    class="sc-hKMtDE keuPlN"
  >
    <div
      class="sc-ftvSNZ cNEgUJ"
      data-testid="recipe-container"
    >
      <div
        class="sc-paqrt ghgffe"
        data-testid="recipe-content"
      >
        <div
          class="sc-jqUWZQ jkOAMd"
        >
          <img
            alt="Crispy Chicken and Rice	with Peas & Arugula Salad"
            data-testid="recipe-image"
            src="//images.ctfassets.net/kk2bw5ojx476/5mFyTozvSoyE0Mqseoos86/fb88f4302cfd184492e548cde11a2555/SKU1479_Hero_077-71d8a07ff8e79abcb0e6c0ebf0f3b69c.jpg"
          />
        </div>
        <div
          class="sc-kDDqET bOCbPU"
        >
          <h1
            class="sc-iqcnOu frMNUQ"
            data-testid="recipe-title"
          >
            Crispy Chicken and Rice	with Peas & Arugula Salad
          </h1>
          <div
            class="sc-crXcXV gXqTwt"
            data-testid="recipe-description"
          >
            <span
              class="markdown"
              data-testid="markdown"
            >
              Crispy chicken skin, tender meat, and rich, tomatoey sauce form a winning trifecta of delicious in this one-pot braise. We spoon it over rice and peas to soak up every last drop of goodness, and serve a tangy arugula salad alongside for a vibrant boost of flavor and color. Dinner is served! Cook, relax, and enjoy!
            </span>
          </div>
          <div
            class="sc-bretMz keDIXw"
            data-testid="recipe-tags-container"
          >
            <span
              class="sc-ewahw kEhVVZ"
            >
              Tags:
            </span>
            <div
              class="sc-ksZauW Jqoyc"
              data-testid="recipe-tags"
            >
              <div
                data-testid="recipe-tag"
              >
                gluten free
              </div>
              <div
                data-testid="recipe-tag"
              >
                healthy
              </div>
            </div>
          </div>
          <div
            class="sc-hAZoWV geqfgz"
            data-testid="recipe-chef"
          >
            <span
              class="sc-ewahw kEhVVZ"
            >
              Chef:
            </span>
            <span>
              Jony Chives
            </span>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
